### PR TITLE
feat: impl oss user restriction

### LINKED ISSUE #3243: Create APIs for OSS user management
### Tasks
- [x] DB changes
- [x] Create APIs for OSS user management
- [x] Test cases

### References
- [Product doc](https://www.notion.so/matterslab/64a78214148645c5a8a09185923dbac8)

### Blockers
- https://github.com/thematters/developer-resource/issues/204

### PR DESCRIPTION
**Pull request by gary02**
- add db migration
- update graphQL schema
-  impl related resolvers
- refactors

resolve #3243

### CODE REVIEW

File: db/migrations/20230406025628_create_user_restriction_table.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+const table = 'user_restriction'
+
+exports.up = async (knex) => {
+  await knex('entity_type').insert({ table })
+  await knex.schema.createTable(table, (t) => {
+    t.bigIncrements('id').primary()
+    t.bigInteger('user_id').unsigned().notNullable()
+    t.enu('type', ['articleHottest', 'articleNewest']).notNullable()
+    t.timestamp('created_at').defaultTo(knex.fn.now())
```
Comment (robertu7) [inline db/migrations/20230406025628_create_user_restriction_table.js:11]: shall we need a `state` (`archived`, `active`) field so we don't need to delete the record?

File: src/connectors/userService.ts [hunk 2]
```diff
@@ -2092,6 +2094,67 @@ export class UserService extends BaseService {
...
+    })
+  }
+
+  updateRestrictions = async (id: string, types: GQLUserRestrictionType[]) => {
+    const olds = (await this.findRestrictions(id)).map(({ type }) => type)
+    const news = [...new Set(types)]
+    const toAdd = news.filter((i) => !olds.includes(i))
+    const toDel = olds.filter((i) => !news.includes(i))
+    await Promise.all(
+      toAdd
```
Comment (robertu7) [inline src/connectors/userService.ts:2118]: can use `[...toAdd.map(...), ...toDel.map(...)]` instead

File: db/migrations/20230406025628_create_user_restriction_table.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+    t.index(['user_id'])
```
Comment (robertu7) [inline db/migrations/20230406025628_create_user_restriction_table.js:13]: also index `type`?

**Review by robertu7 — APPROVED** [review]
🈂️

File: src/connectors/userService.ts (hunk 2 above)
Comment (gary02) [reply src/connectors/userService.ts:2118]: so neat!

File: db/migrations/20230406025628_create_user_restriction_table.js (hunk 3 above)
Comment (gary02) [reply db/migrations/20230406025628_create_user_restriction_table.js:13]: enum column may not benefit too much from indexing ?

File: db/migrations/20230406025628_create_user_restriction_table.js (hunk 1 above)
Comment (gary02) [reply db/migrations/20230406025628_create_user_restriction_table.js:11]: the `state` col will make db operation a little more complex, and according the production requirement by now,  we do not need keep the history of the restriction operation?

Comment (tx0c) [reply db/migrations/20230406025628_create_user_restriction_table.js:11]: some `state` kind of soft delete is still better; maybe some time later we will want to analyze

File: db/migrations/20230406025628_create_user_restriction_table.js (hunk 3 above)
Comment (robertu7) [reply db/migrations/20230406025628_create_user_restriction_table.js:13]: we use filtering by `type` in [recommendation resolvers](https://github.com/thematters/matters-server/pull/3273/files#diff-f090852af5161630e96eabe822e33d6c3dc8aefcc8812c00807235f6a39a4728R45-R50), any performance impact?

Comment (gary02) [reply db/migrations/20230406025628_create_user_restriction_table.js:13]: Indexed enum column first scan index and read, let say 30%, rows row by row from disk,  comparing to batch read all rows from disk in the un-indexed case,   it hard to say which one can be faster.

**Pull request merged**